File: openmetadata-sdk/src/main/java/org/openmetadata/sdk/entities/TestCase.java
package org.openmetadata.sdk.entities;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import org.openmetadata.sdk.client.OpenMetadata;
import org.openmetadata.sdk.client.OpenMetadataClient;
import org.openmetadata.sdk.exceptions.OpenMetadataException;
import org.openmetadata.sdk.models.ListParams;
import org.openmetadata.sdk.models.ListResponse;

/**
 * SDK wrapper for TestCase operations.
 * This class provides static methods for TestCase CRUD operations.
 * It does NOT extend the schema TestCase class to avoid naming conflicts.
 */
public class TestCase {

  private static OpenMetadataClient defaultClient;

  public static void setDefaultClient(OpenMetadataClient client) {
    defaultClient = client;
  }

  private static OpenMetadataClient getClient() {
    if (defaultClient == null) {
      return OpenMetadata.client();
    }
    return defaultClient;
  }

  // Static methods for CRUD operations
  public static org.openmetadata.schema.tests.TestCase create(
      org.openmetadata.schema.api.tests.CreateTestCase createTestCase)
      throws OpenMetadataException {
    return getClient().testCases().create(createTestCase);
  }

  public static org.openmetadata.schema.tests.TestCase retrieve(String id)
      throws OpenMetadataException {
    return getClient().testCases().get(id);
  }

  public static org.openmetadata.schema.tests.TestCase retrieve(String id, String fields)
      throws OpenMetadataException {
    return getClient().testCases().get(id, fields);
  }

  public static org.openmetadata.schema.tests.TestCase retrieve(UUID id)
      throws OpenMetadataException {
    return retrieve(id.toString());
  }

  public static org.openmetadata.schema.tests.TestCase retrieveByName(String name)
      throws OpenMetadataException {
    return getClient().testCases().getByName(name);
  }

  public static org.openmetadata.schema.tests.TestCase retrieveByName(String name, String fields)
      throws OpenMetadataException {
    return getClient().testCases().getByName(name, fields);
  }

  public static org.openmetadata.schema.tests.TestCase update(
      String id, org.openmetadata.schema.tests.TestCase patch) throws OpenMetadataException {
    return getClient().testCases().update(id, patch);
  }

  public static void delete(String id) throws OpenMetadataException {
    getClient().testCases().delete(id);
  }

  public static void delete(UUID id) throws OpenMetadataException {
    getClient().testCases().delete(id);
  }

  public static void delete(String id, boolean recursive, boolean hardDelete)
      throws OpenMetadataException {
    Map<String, String> params = new HashMap<>();
    params.put("recursive", String.valueOf(recursive));
    params.put("hardDelete", String.valueOf(hardDelete));
    getClient().testCases().delete(id, params);
  }

  // Async delete methods
  public static CompletableFuture<Void> deleteAsync(String id) {
    return getClient().testCases().deleteAsync(id);
  }

  public static CompletableFuture<Void> deleteAsync(UUID id) {
    return getClient().testCases().deleteAsync(id);
  }

  public static CompletableFuture<Void> deleteAsync(
      String id, boolean recursive, boolean hardDelete) {
    return CompletableFuture.runAsync(
        () -> {
          try {
            delete(id, recursive, hardDelete);
          } catch (OpenMetadataException e) {
            throw new RuntimeException(e);
          }
        });
  }

  public static ListResponse<org.openmetadata.schema.tests.TestCase> list()
      throws OpenMetadataException {
    return getClient().testCases().list();
  }

  public static ListResponse<org.openmetadata.schema.tests.TestCase> list(ListParams params)
      throws OpenMetadataException {
    return getClient().testCases().list(params);
  }
}

File: ingestion/src/metadata/sdk/entities/testcase.py
"""
Test Case entity implementation for OpenMetadata Python SDK.
"""
from typing import List, Optional

from metadata.generated.schema.api.tests.createTestCase import CreateTestCaseRequest
from metadata.generated.schema.entity.tests.testCase import TestCase as TestCaseEntity
from metadata.sdk.entities.base import BaseEntity


class TestCase(BaseEntity):
    """Test Case entity operations for OpenMetadata SDK"""

    @classmethod
    def entity_type(cls):
        """Return the entity type"""
        return TestCaseEntity

    @classmethod
    def create(cls, request: CreateTestCaseRequest) -> TestCaseEntity:
        """
        Create a new test case.

        Args:
            request: CreateTestCaseRequest with test case details

        Returns:
            Created TestCase entity
        """
        client = cls._get_client()
        return client.create_or_update(request)

    @classmethod
    def retrieve(
        cls, entity_id: str, fields: Optional[List[str]] = None
    ) -> TestCaseEntity:
        """
        Retrieve a test case by ID.

        Args:
            entity_id: TestCase UUID
            fields: Optional list of fields to include

        Returns:
            TestCase entity
        """
        client = cls._get_client()
        return client.get_by_id(
            entity=TestCaseEntity, entity_id=entity_id, fields=fields
        )

    @classmethod
    def retrieve_by_name(
        cls, fqn: str, fields: Optional[List[str]] = None
    ) -> TestCaseEntity:
        """
        Retrieve a test case by fully qualified name.

        Args:
            fqn: Fully qualified name
            fields: Optional list of fields to include

        Returns:
            TestCase entity
        """
        client = cls._get_client()
        return client.get_by_name(entity=TestCaseEntity, fqn=fqn, fields=fields)

    @classmethod
    def update(cls, entity_id: str, entity: TestCaseEntity) -> TestCaseEntity:
        """
        Update a test case (PUT operation).

        Args:
            entity_id: TestCase UUID
            entity: Updated TestCase entity

        Returns:
            Updated TestCase entity
        """
        entity.id = entity_id
        client = cls._get_client()
        return client.create_or_update(entity)

    @classmethod
    def patch(cls, entity_id: str, json_patch: List[dict]) -> TestCaseEntity:
        """
        Patch a test case (PATCH operation).

        Args:
            entity_id: TestCase UUID
            json_patch: JSON patch operations

        Returns:
            Patched TestCase entity
        """
        client = cls._get_client()
        return client.patch(
            entity=TestCaseEntity, entity_id=entity_id, json_patch=json_patch
        )

    @classmethod
    def delete(
        cls, entity_id: str, recursive: bool = False, hard_delete: bool = False
    ) -> None:
        """
        Delete a test case.

        Args:
            entity_id: TestCase UUID
            recursive: Delete recursively
            hard_delete: Hard delete (permanent)
        """
        client = cls._get_client()
        client.delete(
            entity=TestCaseEntity,
            entity_id=entity_id,
            recursive=recursive,
            hard_delete=hard_delete,
        )

    @classmethod
    def list(
        cls,
        fields: Optional[List[str]] = None,
        after: Optional[str] = None,
        before: Optional[str] = None,
        limit: int = 100,
    ) -> List[TestCaseEntity]:
        """
        List test cases.

        Args:
            fields: Fields to include
            after: Pagination cursor
            before: Pagination cursor
            limit: Number of results

        Returns:
            List of TestCase entities
        """
        client = cls._get_client()
        response = client.list_entities(
            entity=TestCaseEntity,
            fields=fields,
            after=after,
            before=before,
            limit=limit,
        )
        return response.entities if response else []
